cmd/linkcheck/links.go

```diff
@@ -54,16 +62,52 @@ func newWalkFunc(invalidLink *bool) filepath.WalkFunc {
 			return err
 		}
 		foundInvalid := false
-		matches := httpRE.FindAllSubmatch(fileBytes, -1)
-		for _, match := range matches {
-			//match[1] should look like docs/devel/api-conventions.md
-			if _, err := os.Stat(path.Join(*repoRoot, string(match[1]))); err != nil {
-				fmt.Fprintf(os.Stderr, "Link is not valid: %s\n", string(match[0]))
+		allURL := xurls.Strict.FindAll(fileBytes, -1)
+		fmt.Fprintf(os.Stdout, "\nChecking file %s\n", filePath)
+	URL:
+		for _, URL := range allURL {
+			// Don't check non http/https URL
+			if !httpOrhttpsReg.Match(URL) {
+				continue
+			}
+			for _, whiteURL := range whiteList {
+				if whiteURL.Match(URL) {
+					continue URL
+				}
+			}
+			// remove the htmlpreview Prefix
+			processedURL := htmlpreviewReg.ReplaceAll(URL, []byte{})
+			retry := 0
+			const maxRetry int = 3
+			backoff := 100
+			for retry < maxRetry {
+				fmt.Fprintf(os.Stdout, "Visiting %s\n", string(processedURL))
+				resp, err := client.Head(string(processedURL))
+				// We only check response code >= 400. We don't care about
+				// errors.
+				if err != nil {
+					break
+				}
+				if resp.StatusCode == 429 {
+					fmt.Fprintf(os.Stderr, "Got %d visiting %s, retry after %d seconds.\n", resp.StatusCode, string(URL), backoff)
```

Comment:
The right way is to read the "Retry-after" header which should tell you the number of seconds you should wait before making another call.